src/server/pkg/watch/watch.go

```diff
@@ -113,62 +94,70 @@ func NewWatcher(ctx context.Context, client *etcd.Client, trimPrefix, prefix str
 			close(eventCh)
 			etcdWatcher.Close()
 		}()
-		for _, etcdKv := range resp.Kvs {
-			eventCh <- &Event{
-				Key:      bytes.TrimPrefix(etcdKv.Key, []byte(trimPrefix)),
-				Value:    etcdKv.Value,
-				Type:     EventPut,
-				Rev:      etcdKv.ModRevision,
-				Ver:      etcdKv.Version,
-				Template: template,
-			}
-		}
 		for {
-			var resp etcd.WatchResponse
-			var ok bool
-			select {
-			case resp, ok = <-rch:
-			case <-done:
-				return nil
+			resp, err := client.Get(ctx, prefix, getOptions...)
```

Comment:
I don't think this is going to give us the correct semantics. When the watcher is recreated, we are going to return a bunch of duplicate put events from the get. Also, I think it is going to drop delete events from revisions between when the watcher errors and when the get occurs. Even if it gets recreated before another revision it is going to miss any deletes from the revision that errored since it will get that revision, not watch it. So, you would end up with rev x (get) -> rev x+1 (watch). What was actually happening that was causing the busy loop? I am assuming that the watcher recreation wasn't blocking, therefore you made us block on the get instead? If so, we should maybe have the watcher recreation in a backoff?

Reply:
So the watch only gets recreated when the previous one encounters an unrecoverable error. The specific error that I was debugging was `ErrCompacted` which happens if things you're watching get compacted before they can get sent over. At that point all you can do is issue a new get to figure out what new revision you should wait for, and there's no risk of duplicate data because the whole point is that etcd already compacted revisions you haven't seen yet, and thus has also compacted all the revisions you have seen. You do have the flipside which is that you've missed events, I'm not sure what we can do about that though, because the data is gone. For errors other than `ErrCompacted` I agree that this will get us duplicate events, that's pretty easy to solve by just filtering out events less then the latest revision we've seen though.

The busy loop was caused by recreating a watcher with the same parameters that the `ErrCompacted` one had been using, which just caused it to error again with the same error. So the get is necessary to find new parameters we can pass to `Watch` that won't error, not to block. I don't think add backoff to the watcher does any good, etcd actually has backoff internally for recoverable errors, so we're only handling non-recoverable errors here, in which case I don't see an option other than doing another get.

Reply:
Okay, I see why that is the only way to proceed when `ErrCompacted` occurs, and why a backoff does not make sense here. I think you would still get duplicate data though because the keys that existed at the compacted revision will be returned by the get after the compaction, so the watcher abstraction we expose would generate put events for them before and after the compaction. It seems like etcd compactions are going to be problematic for the semantics of the clients because several of them depend on delete events (an example that comes to mind is workers watching for a datum set claim lease to expire). If we cannot reliably get delete events, then we have some pretty significant changes to make. The other option is to somehow limit when compactions occur to avoid this issue. I would guess the current system has this problem as is. It seems like the change you mentioned (filtering out the kvs from before the most recently seen revision) should get us close to the same semantics as the current system. This still misses the in between deletes though, where as the prior implementation does not. It seems like it makes sense to special case this to the compaction error.